internal/cosesign1/makedidx509.go

```diff
@@ -0,0 +1,160 @@
+	{x509.ExtKeyUsageCodeSigning, oidExtKeyUsageCodeSigning},
+	{x509.ExtKeyUsageEmailProtection, oidExtKeyUsageEmailProtection},
+	{x509.ExtKeyUsageIPSECEndSystem, oidExtKeyUsageIPSECEndSystem},
+	{x509.ExtKeyUsageIPSECTunnel, oidExtKeyUsageIPSECTunnel},
+	{x509.ExtKeyUsageIPSECUser, oidExtKeyUsageIPSECUser},
+	{x509.ExtKeyUsageTimeStamping, oidExtKeyUsageTimeStamping},
+	{x509.ExtKeyUsageOCSPSigning, oidExtKeyUsageOCSPSigning},
+	{x509.ExtKeyUsageMicrosoftServerGatedCrypto, oidExtKeyUsageMicrosoftServerGatedCrypto},
+	{x509.ExtKeyUsageNetscapeServerGatedCrypto, oidExtKeyUsageNetscapeServerGatedCrypto},
+	{x509.ExtKeyUsageMicrosoftCommercialCodeSigning, oidExtKeyUsageMicrosoftCommercialCodeSigning},
+	{x509.ExtKeyUsageMicrosoftKernelCodeSigning, oidExtKeyUsageMicrosoftKernelCodeSigning},
+}
+
+func MakeDidX509(fingerprintAlgorithm string, fingerprintIndex int, chainFilename string, didPolicy string, verbose bool) (string, error) {
+	if fingerprintAlgorithm != "sha256" {
+		return "", fmt.Errorf("unsupported hash algorithm")
+	}
+
+	if fingerprintIndex == 0 {
+		return "", fmt.Errorf("fingerprint index must be >= 1")
+	}
+
+	var chainPEM = string(ReadBlob(chainFilename))
+	chain, err := parsePemChain(chainPEM)
+
+	if err != nil {
+		return "", err
+	}
+
+	if fingerprintIndex > len(chain) {
+		return "", fmt.Errorf("signer index out of bounds")
+	}
+
+	signerCert := chain[fingerprintIndex]
+	var hash = sha256.Sum256(signerCert.Raw)
+	fingerprint := base64.RawURLEncoding.EncodeToString(hash[:])
+
+	var policyTokens []string
+	didPolicyUpper := strings.ToUpper(didPolicy)
+	switch didPolicyUpper {
+	case "CN":
+		{
+			policyTokens = append(policyTokens, "subject", "CN", chain[0].Subject.CommonName)
+		}
+	case "EKU":
+		{
+			// Note: In general there may be many predefined and not predefined key usages.
+			// We pick the first non-predefined one, or, if there are none, the first predefined one.
+			// Others must be specified manually (as in custom, next branch).
+
+			// non-predefined
+			if len(chain[0].UnknownExtKeyUsage) > 0 {
+				policyTokens = append(policyTokens, "eku", chain[0].UnknownExtKeyUsage[0].String())
+			} else if len(chain[0].ExtKeyUsage) > 0 {
+				// predefined
+				policyTokens = append(policyTokens, "eku", extKeyUsageOIDs[chain[0].ExtKeyUsage[0]].oid.String())
+			}
+		}
+	default:
+		{
+			// Custom policies
+			policyTokens = strings.Split(didPolicy, ":")
+		}
+	}
+
+	if len(policyTokens) == 0 {
+		return "", fmt.Errorf("invalid policy")
+	}
+
+	for i := 0; i < len(policyTokens); i++ {
+		policyName := policyTokens[i]
+		switch policyName {
+		case "subject":
+			{
+				i += 2
+				if i >= len(policyTokens) {
+					return "", fmt.Errorf("invalid '%s' policy", policyName)
+				}
+				policyTokens[i] = url.PathEscape(policyTokens[i])
+			}
+		default:
+			{
+			}
+		}
+	}
+
+	r := "did:x509:0:" + fingerprintAlgorithm + ":" + fingerprint + "::" + strings.Join(policyTokens, ":")
```

Comment:
nit: define a format const for this.

Reply:
What do you mean?

Reply:
something like:
```
const issuerFmt = "did:x509:0:%s:%s::%s"
```
and then here you'd do
```
r := fmt.Sprintf(issuerFmt, fingerprintAlgorithm, fingerprint, strings.Join(policyTokens, ":")
```

Reply:
https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

What happens when I change "issuerFmt"? Say it turns out we need the "0" to be a number. Now I am at risk at thinking it is done. It still compiles. But really, I have to change the match place it is used. Perhaps the linter will catch it, perhaps some test.

How is that better?

Looking at the code as it stands it is very obvious what is going on.